### PR TITLE
[JUJU-804] Remove facade reflection during creation.

### PR DESCRIPTION
**Pull request by SimonRichardson**
Using reflection for every facade construction was a flex, that was dumb. 
Instead, we should be boring and create facades using functions that are 
typed by the compiler, so if we make any mistakes we don’t have to wait 
till runtime to figure out what we broke!

The following also introduces the concept of inversion of control when
registering a facade. Instead of the all facades file knowing about each
and every version of the facade, we invert that. Now we ask the package
what it would like to register and it is up to that package what it's
registering. We no longer require exported functions to be randomly
exported. It is more considered. This reduces the footprint of facades
public API. The downside for now, that will need to be improved is the
usage of export_test.go files. A lot of the tests are functional/behavioural 
tests, when they should be unit tests (both have their place).

The advantage to doing all this work, is that shimming a API facade layer
in front will likely be a lot simpler as reflection isn't so extensively
used.

## QA steps

Ensure when rebuilding the schema we didn't break anything. That
outputs the latest facades in json format and we can diff that output
against the tip of 2.9 to see that indeed nothing changed.

```sh
$ make rebuild-schema
$ diff <2.9 schema.json> <new schema.json>
```

Run some tests to ensure we didn't break anything.

```sh
$ juju bootstrap lxd test
$ juju deploy ubuntu
$ juju status
```

### CODE REVIEW

File: apiserver/facade/registry.go [hunk 1]
```diff
@@ -185,75 +179,3 @@ func (f *Registry) Discard(name string, version int) {
...
-		in := []reflect.Value{reflect.ValueOf(context)}
-		out := funcValue.Call(in)
-		if out[1].Interface() != nil {
-			err := out[1].Interface().(error)
-			return nil, err
-		}
-		return out[0].Interface(), nil
-	}
-	return wrapped, funcValue.Type().Out(0), nil
-}
```
Comment (SimonRichardson) [inline apiserver/facade/registry.go:259]: Removed the runtime checking. It's gone.

File: apiserver/facades/agent/agent/register.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/common/cloudspec"
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/apiserver/facade"
+)
+
+// Registry describes the API facades exposed by some API server.
+type Registry interface {
```
Comment (manadart) [inline apiserver/facades/agent/agent/register.go:18]: I know we should declare these close to usage, but I think it's reasonable to have this up in the `apiserver/facade` or `apiserver/facades` package for brevity.

File: apiserver/allfacades.go [hunk 3]
```diff
@@ -117,306 +113,131 @@ import (
...
-	reg("Client", 4, client.NewFacadeV4)
-	reg("Client", 5, client.NewFacade)
-	reg("Cloud", 1, cloud.NewFacadeV1)
-	reg("Cloud", 2, cloud.NewFacadeV2) // adds AddCloud, AddCredentials, CredentialContents, RemoveClouds
-	reg("Cloud", 3, cloud.NewFacadeV3) // changes signature of UpdateCredentials, adds ModifyCloudAccess
-	reg("Cloud", 4, cloud.NewFacadeV4) // adds UpdateCloud
-	reg("Cloud", 5, cloud.NewFacadeV5) // Removes DefaultCloud, handles config in AddCloud
-	reg("Cloud", 6, cloud.NewFacadeV6) // Adds validity to CredentialContent, force for AddCloud
-	reg("Cloud", 7, cloud.NewFacadeV7) // Do not set error if forcing credential update.
+	action.Register(registry)
```
Comment (manadart) [inline apiserver/allfacades.go:116]: Much better.

**Review by manadart — APPROVED** [review]
This makes me happy.

I also successfully executed model migration and CMR.

File: apiserver/allfacades.go (hunk 3 above)
Comment (SimonRichardson) [reply apiserver/allfacades.go:116]: Agreed

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**